src/Microsoft.OData.Core/JsonLight/ODataJsonLightResourceDeserializer.cs

```diff
@@ -1061,136 +1062,2239 @@ internal ODataJsonLightReaderNestedInfo ReadPropertyWithoutValue(IODataJsonLight
+            if (propertyParsingResult == PropertyParsingResult.ODataInstanceAnnotation)
+            {
+                resourceState.PropertyAndAnnotationCollector.AddODataScopeAnnotation(annotationName, value);
+            }
+            else
+            {
+                resourceState.PropertyAndAnnotationCollector.AddCustomScopeAnnotation(annotationName, value);
+            }
+
+            return value;
+        }
+
+        /// <summary>
+        /// Asynchronously reads the resource set instance annotations for a top-level resource set.
+        /// </summary>
+        /// <param name="resourceSet">The <see cref="ODataResourceSet"/> to read the instance annotations for.</param>
+        /// <param name="propertyAndAnnotationCollector">The duplicate property names checker for the top-level scope.</param>
+        /// <param name="forResourceSetStart">true when parsing the instance annotations before the resource set property;
+        /// false when parsing the instance annotations after the resource set property.</param>
+        /// <param name="readAllResourceSetProperties">true if we should scan ahead for the annotations and ignore the actual data properties (used with
+        /// the reordering reader); otherwise false.</param>
+        /// <returns>A task that represents the asynchronous read operation.</returns>
+        internal async Task ReadTopLevelResourceSetAnnotationsAsync(
+            ODataResourceSetBase resourceSet,
+            PropertyAndAnnotationCollector propertyAndAnnotationCollector,
+            bool forResourceSetStart,
+            bool readAllResourceSetProperties)
+        {
+            Debug.Assert(resourceSet != null, "resourceSet != null");
+            Debug.Assert(propertyAndAnnotationCollector != null, "propertyAndAnnotationCollector != null");
+            this.JsonReader.AssertNotBuffering();
+
+            bool buffering = false;
+            try
+            {
+                while (this.JsonReader.NodeType == JsonNodeType.Property)
+                {
+                    bool foundValueProperty = false;
+
+                    if (!forResourceSetStart && readAllResourceSetProperties)
+                    {
+                        // If this is not called for reading ResourceSetStart and we already scanned ahead and processed all resource set properties, we already checked for duplicate property names.
+                        // Use an empty duplicate property name checker since this.ParsePropertyAsync() read through the same property annotation of instance annotations again.
+                        propertyAndAnnotationCollector = new PropertyAndAnnotationCollector(false);
+                    }
+
+                    await this.ProcessPropertyAsync(
+                        propertyAndAnnotationCollector,
+                        this.ReadTypePropertyAnnotationValueAsync,
+                        async (propertyParseResult, propertyName) =>

```

Comment:
To reduce indentation we could move this to another function within the class instead of the lambda. This would reduce the captured arguments e.g. resourceSet  in 1629 which would reduce the number of allocations.

Reply:
To be handled in a separate PR for fixing issues surrounding the extra allocations occasioned by capturing of variables